content/courses/Ada_For_The_C_Embedded_Developer/chapters/08_Performance.rst

```diff
@@ -29,10 +29,103 @@ suspects and their resolution.
+| ``-O2``     | Full optimization.                                           |
++-------------+--------------------------------------------------------------+
+| ``-O3``     | Same optimization level as for ``-O2``. In addition, further |
+|             | optimization strategies, such as aggressive automatic        |
+|             | inlining and vectorization.                                  |
++-------------+--------------------------------------------------------------+
+
+Note that the highest the level, the slowest will be the compilation time. For
+fast compilation during development phase, unless you're working on
+benchmarking algorithms, using ``-O0`` is probably a good idea.
+
+In addition to the levels presented above, GNAT also has the ``-Os`` switch,
+which allows for optimizing code and data usage.
+
+Inlining
+~~~~~~~~
+
+As we've seen in the previous section, the highest optimization level (``-O3``)
+tries to automatic inline subprograms whenever possible. There might be
+situations, however, where this doesn't happen automatically. In those cases,
+manual inlining may help.
+
+In order to inline subprograms, we can use the :ada:`Inline` aspect. Let's
+reuse an example from a previous chapter and inline the :ada:`Average`
+function:
+
+.. code:: ada run_button project=Courses.Ada_For_C_Embedded_Dev.Performance.Inlining
+
+    package Float_Arrays is
+
+       type Float_Array is array (Positive range <>) of Float;
+
+       function Average (Data : Float_Array) return Float
+         with Inline;
+
+    end Float_Arrays;
+
+    package body Float_Arrays is
+
+       function Average (Data : Float_Array) return Float is
+          Total : Float := 0.0;
+       begin
+          for Value of Data loop
+             Total := Total + Value;
+          end loop;
+          return Total / Float (Data'Length);
+       end Average;
+
+    end Float_Arrays;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    with Float_Arrays; use Float_Arrays;
+
+    procedure Compute_Average is
+       Values        : constant Float_Array := (10.0, 11.0, 12.0, 13.0);
+       Average_Value : Float;
+    begin
+       Average_Value := Average (Values);
+       Put_Line ("Average = " & Float'Image (Average_Value));
+    end Compute_Average;
+
+When compiling this example, the GNAT will inline :ada:`Average` in the
+:ada:`Compute_Average` procedure.
+
+In addition to this aspect, we need to set the optimization level to at least
+``-O1`` and use the ``-gnatn`` switch, which instructs the compiler to take the
+:ada:`Inline` aspect into account.
+
+Note, however, that the :ada:`Inline` aspect is just a *recommendation* to the
+compiler. Sometimes, the compiler might not be able to follow this
+recommendation, so it won't inline the subprogram. If we want to ensure that a
+subprogram is always inlined or else get a compilation error when it isn't, we
+need to use the :ada:`Inline_Always` aspect instead of :ada:`Inline`.
```

Comment:
I was told (by Robert, who wrote the compiler) not to use Inline_Always for that reason because it was intended only for cases in which the functionality would be incorrect absent the inlining. So use it on procedures that insert assembly instructions that only make sense inline, such as memory barriers, but nothing else.
With the aspect/pragma Inline and and the necessary switches we do get a warning when inline cannot be achieved.